clang/test/SemaSYCL/restrict-recursion4.cpp

```diff
@@ -24,7 +24,7 @@ void *operator new(size_t);
 void usage2(myFuncDef functionPtr) {
```

Comment:
Could it be possible to add a test like
```c++
void kernel3(void) {
  if constexpr (false)
    kernel3();
  if (false)
    kernel3();
}
```
and verify it does not fire?

Reply:
Hi, @keryell 
I am afraid current implementation can't handle this as we are analyzing the call graph to recognize all recursions and don't know whether the recursive call is in a "dead" branch.
Thanks very much.

Reply:
I think it's reasonable to not diagnose recursion in dead branches, but we should still have test coverage (with comments) that show we've explicitly considered this case.

Reply:
This comment is still unaddressed, btw.

Reply:
@AaronBallman :Note that my patch adds a test for this.  See: https://github.com/intel/llvm/pull/3475